### PR TITLE
[4.0.x] Pagination supports optional aliases and Phalcon\Mvc\Model\Query\Builder supports bind params

### PR DESCRIPTION
**Pull request by CameronHall**
Hello!

* Type: new feature
* Link to issue: https://github.com/phalcon/cphalcon/issues/10957, https://github.com/phalcon/cphalcon/issues/13368
* Link to existing PR: https://github.com/phalcon/cphalcon/pull/10985

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change: 
- Added `Phalcon\Paginator\RepositoryInterface` for repository the current state of `paginator` and also optional sets the aliases for properties repository. 
- Added bind support to `Phalcon\Mvc\Model\Query\Builder`. The Query Builder has the same methods as `Phalcon\Mvc\Model\Query`; `getBindParams`, `setBindParams`, `getBindTypes` and `setBindTypes`. [#13368](https://github.com/phalcon/cphalcon/issues/13368)

It is worth noting I made a few changes between the original PR https://github.com/phalcon/cphalcon/pull/10985 and this one. I've updated all the property names to what they were changes to in https://github.com/phalcon/cphalcon/issues/13492. I also turfed the `totalPages` property as it was made redundant by the pre-existing property `last`. 

All of the pagination adapter constructors received a spring clean as well.

Thanks

### CODE REVIEW

File: phalcon/paginator/adapter/querybuilder.zep [hunk 1]
```diff
@@ -68,30 +64,24 @@ class QueryBuilder extends Adapter
...
-			throw new Exception("Parameter 'limit' is required");
-		}
-
 		if fetch columns, config["columns"] {
 		    let this->_columns = columns;
 		}
+		
+		if !isset config["limit"] {
+			throw new Exception("Parameter 'limit' is required");
+		}
```
Comment (niden) [inline phalcon/paginator/adapter/querybuilder.zep]: Move this after the `fetch` for the `config["builder"]`. We want to throw an exception as early as possible and avoid extra processing - if the data is incorrect

File: tests/integration/Paginator/Adapter/QueryBuilderCest.php [hunk 2]
```diff
@@ -136,33 +138,64 @@ public function testIssue12111WithoutGroup(IntegrationTester $I)
...
-                        "page"    => 2
-                    ]
-                ))->paginate();
+        $this->setDiMysql();
+        $modelsManager = $this->getService('modelsManager');
 
-                expect($paginate->last)->equals(2);
-                expect($paginate->total_items)->equals(2);
-            }
+        $di = $this->getDI();
```
Comment (niden) [inline tests/integration/Paginator/Adapter/QueryBuilderCest.php]: `$di = $this->getDI()` => `$this->container`

So you can do `$this->container->set('dbTwo', $this->container->get('db'));`

File: tests/integration/Paginator/Adapter/RepositoryCest.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
+<?php
```
Comment (niden) [inline tests/integration/Paginator/Adapter/RepositoryCest.php]: Will need to refactor this a bit after we add it to the main repo. I will generate test stubs to split all the methods and make tests shorter. a todo for later.

**Review by niden — CHANGES_REQUESTED** [review]

**Comment by niden** [comment]
Looking good!

File: tests/integration/Paginator/Adapter/RepositoryCest.php (hunk 3 above)
Comment (CameronHall) [reply tests/integration/Paginator/Adapter/RepositoryCest.php]: No worries. In future, would you rather we generate the stubs when adding new classes? If so, how do we do that?

Comment (niden) [reply tests/integration/Paginator/Adapter/RepositoryCest.php]: Those are generated with `zephir stubs`

**Review by niden — APPROVED** [review]

**Comment by niden** [comment]
@CameronHall Thank you!

File: .gitignore [hunk 4]
```diff
@@ -54,3 +54,5 @@ php_test_results_*.txt
 docker-compose.yml
 build/gccarch
 tests/_cache
+.zephir/
+.temp/
```
Comment (sergeyklay) [inline .gitignore]: @CameronHall Zephir >= 0.11 uses `.zephir` and Zephir < 0.11` .temp`

File: phalcon/mvc/model/query/builderinterface.zep [hunk 5]
```diff
@@ -202,6 +202,25 @@ interface BuilderInterface
 	/**
 	 * Sets an OFFSET clause
 	 */
-	 public function offset(int offset) -> <BuilderInterface>;
+	public function offset(int offset) -> <BuilderInterface>;
```
Comment (sergeyklay) [inline phalcon/mvc/model/query/builderinterface.zep]: Did you use spaces instead of tabs?

File: .gitignore (hunk 4 above)
Comment (CameronHall) [reply .gitignore]: @sergeyklay I'll remove `.temp` in the next PR.

File: phalcon/mvc/model/query/builderinterface.zep (hunk 5 above)
Comment (CameronHall) [reply phalcon/mvc/model/query/builderinterface.zep]: I hope not. It should have been picked up by `phpcs` if I did. If I don't fix it in my next PR assume it's a non-issue.

Comment (sergeyklay) [reply phalcon/mvc/model/query/builderinterface.zep]: @CameronHall phpcs know nothing about *.zep files

Comment (CameronHall) [reply phalcon/mvc/model/query/builderinterface.zep]: RIP. I'll fix in my next PR which will be addressing #13552.

Comment (CameronHall) [reply phalcon/mvc/model/query/builderinterface.zep]: @sergeyklay looks like I removed the space in the commit so it already existed I suppose 👍

Comment (sergeyklay) [reply phalcon/mvc/model/query/builderinterface.zep]: @CameronHall :+1:  Thank you

**Pull request merged**